### DISCUSSION TITLE
[Proposal]: shorthand syntax for `IEnumerable`

### DISCUSSION THREAD

**Discussion by dme-compunet** [dotnet/csharplang#6835] [Language Ideas]
Proposal:

``` csharp
public int<> SomeMethod()
{
    yield return 1;
    yield return 2;
}
```

Instead of:

``` csharp
public IEnumerable<int> SomeMethod()
{
    yield return 1;
    yield return 2;
}
```

It shortens typing and improves looks.
This will make me and I believe many other developers have more fun when developing with `IEnumerable`.

**Comment by theunrepentantgeek** [comment]
The syntax `int<>` looks like you're declaring a generic subtype of **int**, and I don't know what that would even mean. 

> shortens typing

This isn't a goal of C#. 

Being explicit is good. 
Being needlessly cryptic is not.

**Reply by HaloFour** [reply]
Stupid trivia, but in the research language [Cω (C Omega)](https://en.wikipedia.org/wiki/C%CF%89), which had iterators before C# did, the syntax to declare an iterator was:

https://www.99-bottles-of-beer.net/language-c&sharp969;-(c-omega)-1855.html (geez, that was 15 years ago)

```cs
public int* SomeMethod() {
    yield return 1;
    yield return 2;
}
```

**Comment by svick** [comment]
Personally, I think this makes the code much less clear. And while `IEnumerable<T>` is somewhat special in the language, there are many other collection types, and I'm not certain `IEnumerable<T>` deserves special syntax. (In fact, if I were creating a new language, I probably wouldn't even give special syntax to arrays.)

Also, this syntax is ambigous: `typeof(ValueTuple<>)` is already valid C# code and this proposal would add a second meaning.

And one more thing: I don't think C# is meant to be "fun".

**Comment by dme-compunet** [comment]
Thanks all, you gave me what to think about concerning the goals and ideals of the language.

**Reply by CyrusNajmabadi** [reply]
You're welcome!

**Comment by alrz** [comment]
Using https://github.com/dotnet/csharplang/discussions/90 you'll be able to do:
```cs
global using seq<T> = System.Generic.IEnumerable<T>;
```
(That's what F# uses for the name)
